File: ois/ois20/round3/ois11_round3_solutions.tar(1)/fairgame.py
#!/usr/bin/env pypy3
# NOTE: it is recommended to use this even if you don't understand the following code.


# input data
N, K, M, P, Q = map(int, input().strip().split())

# insert your code here
A = []
B = []
A.append((0, Q-P))

for i in range(1, N+1):
    while A and A[0][0] + K < i:
        A = A[1:]
    while B and B[0][0] + K < i:
        B = B[1:]

    best = -10**9
    if A:
        best = max(best, -A[0][1] - (i&1)*M)
    if B:
        best = max(best, -B[0][1] - (1 - (i&1))*M)

    if i&1:
        while B and B[-1][1] >= best:
            B = B[:-1]
        B.append((i, best))
    else:
        while A and A[-1][1] >= best:
            A = A[:-1]
        A.append((i, best))

print(B[-1][1] if N&1 else A[-1][1])  # print the result

File: ois/ois20/round3/ois11_round3_solutions.tar(1)/fairgame.cpp
/*
 * This template is valid both in C and in C++,
 * so you can expand it with code from both languages.
 * NOTE: it is recommended to use this even if you don't
 * understand the following code.
 */

#include <stdio.h>
#include <assert.h>
#include <deque>
#include <vector>
using namespace std;

typedef long long ll;

const ll INF = 1'000'000'000'000'000'000L;

// input data
int N, K, M, P, Q;

int main() {
//  uncomment the following lines if you want to read/write from files
//  freopen("input.txt", "r", stdin);
//  freopen("output.txt", "w", stdout);

    assert(5 == scanf("%d%d%d%d%d", &N, &K, &M, &P, &Q));

    deque<pair<int, ll>> A, B;
    A.emplace_back(0, Q-P);
    for (int i = 1; i <= N; i++) {
        while (!A.empty() && A.front().first+K < i) A.pop_front();
        while (!B.empty() && B.front().first+K < i) B.pop_front();

        ll best = -INF;
        if (!A.empty()) {
            best = max(best, -A.front().second - (i&1) * M);
        }

        if (!B.empty()) {
            best = max(best, -B.front().second - (1 - (i&1)) * M);
        }

        if (i&1) {
            while (!B.empty() && B.back().second >= best) B.pop_back();
            B.emplace_back(i, best);
        } else {
            while (!A.empty() && A.back().second >= best) A.pop_back();
            A.emplace_back(i, best);
        }
    }

    // insert your code here

    printf("%lld\n", ((N&1) ? B.back().second : A.back().second)); // print the result
    return 0;
}
